### ISSUE TITLE
Provide temporary fix for the "Cannot find trait Hash in this scope" issue

### ISSUE THREAD

**Issue by ironcev** [FuelLabs/sway#5129]
_(Adapted from our discussion)_

When calling one of the hash functions (such as `sha256`) on a type, error is thrown with span inside the standard library saying that the `Hash` trait is out of scope.

![image](https://github.com/FuelLabs/sway/assets/4142833/8b6a4bfa-b8e0-4948-a4b8-ab82a1a1661a)

What the error really means is that the impl for that particular type is not in scope at the call site. To fix the issue one needs to `use std::hash::Hash`.

The underlying issue in Sway is that we don't have a compile step to collect all the impls for a trait, so when you use one it has to be manually imported. And the only way to import impls is by importing the trait they relate to if and only if the impls are in the same module as the trait definition.

Considering how often the issue appears and how confusing the error message is, we want to provide a temporary solution before implementing collecting impls for a trait.

Potential temporary fixes: 
1. Putting `Hash` (and hence the impls) in the prelude. Cons: Bloating the default namespace is bad practice in general. Removing things from prelude later on is a breaking change. Programmers still need to `use std::hash::*`.
1. Provide helpful error message that will point to the solution. We have to see how much effort would it be to come up with a good heuristic and how helpful the message will really be.

**Comment by bitzoic** [comment]
This should resolve the example seen in https://github.com/FuelLabs/sway/issues/5084

**Comment by ironcev** [comment]
> This should resolve the example seen in https://github.com/FuelLabs/sway/issues/5084

Unfortunately not :disappointed: We want to have the traits imported automatically as you mentioned in #5084 but it will take a while till we get there on the compiler side. In the meanwhile we want at least to place the error message at the call site and give the right hint how to solve the issue. That's the direction (No. 2 above) that I am exploring right now.

**Comment by ironcev** [comment]
# TL;DR

To actually solve the issue with traits in trait constraints we will need to:
- autocollect all trait impls for a type
- make `TraitConstraint` aware of exact trait types (`TyTraitConstraint` ?)
- separate `TyTraitDecl` formation from insertion into namespace

# Conclusion

The second option was possbile even without any heuristics. We can precisely identify the issue in a general case, and provide a detailed guideline to programmers how to solve it.

Regarding actually fixing the issue (e.g., #5084) here is a short input on three things we need to do.

## Collecting all trait impls for a type

We do need the mechanism to collect all the trait impls for a type. Interestingly, in all the reported cases of the `Hash` trait related issues, this was not the cause of the issue, because the impls of a type were actually collected and available in the `TraitMap`. But in a general case, when the impl is not located in the same module as the type, we need the mechanisam to collect all the impls.

In reported cases, the issue was that a `TyTraitDecl`s of `TraitConstraint`s are needed during the, e.g, function application or struct instantiation and they are fetched from the `ctx.namespace`. Which means, they need to be imported into the namespace at the call site, which is exactly the manual fix the programmers are forced to do now.

In general, error messages saying that a trait impl cannot be found e.g. when instantiating a struct that has trait constraints, are often wrong, because the impl is actually collected, but the `TyTraitDecl` cannot be found in the `ctx.namespace`.

Locating `TyTraitDecl`s in the `ctx.namespace` on the call site is, of course, conceptually wrong. The current issues we have in getting it from the `DeclEngine` are explained below.

## Making `TraitConstraint` aware of exact trait types

`TraitConstraint`s are created during the parse tree conversion and thus lack propert type information about the trait. A trait in a `TraitConstraint` is "identified" only by its name/ident/suffix and as such ambiguous. Even worse, if an alias is used for a trait at the function declaration site, we even do not have the exact trait name.

The fact that we identify the trait only by it's name (and at the moment actually treat an attempt to use fully qulified name as a compile error) leads to the following issues:

- Obviously, a trait constraint cannot be a combination of different traits of the same name (unless aliases are used, which lead to other problems described below).
- At the call site, the actual trait needs to be resolved which is done by `ctx.namespace.resolve_call_path`. This means, the caller is forced to import at the call site exactly the same trait definition that was used at the definition site. If, by chance, a different trait of the same name is imported at the call site we will end up, depending on concrete circumstance, in various compiler errors, internal compiler error being the most probable but not the only one.
- If an alias is used for a trait at the definition site, the alias name will be taken over in the `TraitConstraint`. This means, at the call site, exactly the same trait alias needs to be used that was used at the definition site.

## Separating `TyTraitDecl` formation from insertion into namespace

Even if we assume that the trait names provided in `TraitConstraint`s are unique and try to get their `TyTraitDecl`s from the `DeclEngine` we will still have an issue.

Essentially, the `TyTraitDecl`s will be available in the `DeclEngine`, and although exactly the same as the one that would be gotten from the `ctx.namespace` (if imported), the rest of the compiler chain does not work properly. E.g., the IR generation will complain that the "Method XZY is a trait method dummy and was not properly replaced." This means that e.g. the calculation of the interfaces surface, replacement of trait dummies etc. currently relies on the import into namespace even on the caller side (the traits are already properly imported into namespaces at the definiton side).
